File: problems/树/113.路径总和 II.py
#!/usr/bin/env python3
"""
113. 路径总和 II
"""
from typing import List, Optional
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def path_sum(root: Optional[TreeNode], target_sum: int) -> List[List[int]]:
    result = []
    path = []
    def dfs(node: Optional[TreeNode], remaining: int):
        if not node: return
        path.append(node.val)
        if not node.left and not node.right and remaining == node.val:
            result.append(path[:])
        dfs(node.left, remaining - node.val)
        dfs(node.right, remaining - node.val)
        path.pop()
    dfs(root, target_sum)
    return result

if __name__ == "__main__":
    root1 = TreeNode(5, TreeNode(4, TreeNode(11, TreeNode(7), TreeNode(2)), None),
                    TreeNode(8, TreeNode(13), TreeNode(4, None, TreeNode(1))))
    print(path_sum(root1, 22))  # [[5,4,11,2], [5,8,4,1]]
    print(path_sum(TreeNode(1), 1))  # [[1]]
    print(path_sum(None, 0))  # []

File: problems/树/113.路径总和 II.cpp
#include <iostream>
#include <vector>
#include <functional>
using namespace std;
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
    vector<vector<int>> result;
    vector<int> path;
    function<void(TreeNode*, int)> dfs = [&](TreeNode* node, int rem) {
        if (!node) return;
        path.push_back(node->val);
        if (!node->left && !node->right && rem == node->val) result.push_back(path);
        dfs(node->left, rem - node->val);
        dfs(node->right, rem - node->val);
        path.pop_back();
    };
    dfs(root, targetSum);
    return result;
}
int main() {
    TreeNode* root = new TreeNode(5);
    root->left  = new TreeNode(4);
    root->right = new TreeNode(8);
    root->left->left  = new TreeNode(11);
    root->right->left  = new TreeNode(13);
    root->right->right = new TreeNode(4);
    root->left->left->left  = new TreeNode(7);
    root->left->left->right = new TreeNode(2);
    root->right->right->right = new TreeNode(1);
    auto r = pathSum(root, 22);
    for (auto& p : r) { for (int v : p) cout << v << " "; cout << endl; }
    // 5 4 11 2 / 5 8 4 1
    r = pathSum(new TreeNode(1), 1);
    for (int v : r[0]) cout << v << " "; cout << endl;  // 1
    cout << pathSum(nullptr, 0).size() << endl;  // 0
    return 0;
}
